src/TesApi.Web/Storage/TerraStorageAccessProvider.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using TesApi.Web.Management.Clients;
+using TesApi.Web.Management.Configuration;
+using TesApi.Web.Management.Models.Terra;
+
+namespace TesApi.Web.Storage
+{
+    /// <summary>
+    /// Provides methods for blob storage access by using local path references in form of /storageaccount/container/blobpath
+    /// </summary>
+    public class TerraStorageAccessProvider : StorageAccessProvider
+    {
+        private readonly TerraOptions terraOptions;
+        private readonly TerraWsmApiClient terraWsmApiClient;
+        private const int FirstSegment = 0;
+        private const int SecondSegment = 1;
+        private const string SasTokenPermissions = "racw";
+
+        /// <summary>
+        /// Provides methods for blob storage access by using local path references in form of /storageaccount/container/blobpath
+        /// for Terra
+        /// </summary>
+        /// <param name="logger">Logger <see cref="ILogger"/></param>
+        /// <param name="terraOptions"><see cref="TerraOptions"/></param>
+        /// <param name="azureProxy">Azure proxy <see cref="IAzureProxy"/></param>
+        /// <param name="terraWsmApiClient"><see cref="terraWsmApiClient"/></param>
+        public TerraStorageAccessProvider(ILogger<TerraStorageAccessProvider> logger,
+            IOptions<TerraOptions> terraOptions, IAzureProxy azureProxy, TerraWsmApiClient terraWsmApiClient) : base(
+            logger, azureProxy)
+        {
+            this.terraWsmApiClient = terraWsmApiClient;
+            ArgumentNullException.ThrowIfNull(terraOptions);
+
+            this.terraOptions = terraOptions.Value;
+
+        }
+
+        /// <inheritdoc />
+        public override Task<bool> IsPublicHttpUrlAsync(string uriString)
+        {
+            //TODO: check if this assumption is correct
+            //For Terra, if it is a Http/Https Url then it is public.
+            return Task.FromResult(TryParseHttpUrlFromInput(uriString, out _));
+
+        }
+
+        /// <inheritdoc />
+        public override async Task<string> MapLocalPathToSasUrlAsync(string path, bool getContainerSas = false)
```

Comment:
@BMurri and @MattMcL4475 can you provide feedback on this logic here? For Terra the expectation is that inputs will be called using the code below. Which means the storage account and container name will be always the two first segments of the path. 

```scala
 val blobSegment = ".blob.core.windows.net"
  def transformBlobString(s: String): String = if (s.contains(blobSegment)) {
    s.replaceFirst("https:/", "").replaceFirst(blobSegment, "")
  } else s
```

Reply:
A `TesInput` (and for that matter, a `TesOutput`) has a `Path` property that is always executor-local (and is always a file-spec). It also has a `Url` property which represents the (possible) source (for `TesInput`) and destination (for `TesOutput`), which is the property that needs to possibly be converted to a URI and possibly need to have a SAS appended.


Most of the time `MapLocalPathToSasUrlAsync()` operates on local paths. Occasionally, as a convenience to the user, it is used for "URL"s that are expressed as if they were a local path, but either are pointing to a `[default-storage]/container/blob-path` or `/storage-account/container/blob-path` path-spec-like expression.